lib/ulid/decoder.ex

```diff
@@ -0,0 +1,77 @@
+defmodule Ulid.Decoder do
+  @moduledoc """
+  Handles the decoding of text Ulids.
+  """
+
+  @doc """
+  Encodes a Ulid binary to its textual form.
+
+  ## Examples
+
+      iex> ulid = "01C9S7RTEBG8NJTYT689ENQD2X"
+      iex> Ulid.Decoder.decode(ulid)
+      {:ok, <<1, 98, 114, 124, 105, 203, 130, 43, 45, 123, 70, 66, 93, 91, 180, 93>>}
+  """
+  @spec decode(Ulid.t()) :: {:ok, Ulid.raw()} | :error
+  def decode(<<_::unsigned-size(208)>> = text) do
+    decode_bytes(text, <<>>)
+  catch
+    :error -> :error
+  else
+    binary -> {:ok, binary}
+  end
+
+  ## PRIVATE FUNCTIONS
+
+  defp decode_bytes(<<>>, acc) do
+    shift = 8 - rem(bit_size(acc), 8)
+    shift_bytes(<<0::size(shift), acc::bitstring>>)
+  end
+
+  defp decode_bytes(<<codepoint::unsigned>>, <<>>) do
+    index = i(codepoint)
+    <<0::3, index::unsigned-size(5)>>
+  end
+
+  defp decode_bytes(<<codepoint::unsigned, rest::bitstring>>, acc) do
+    index = i(codepoint)
+    decode_bytes(rest, <<acc::bitstring, index::unsigned-size(5)>>)
+  end
+
+  defp shift_bytes(<<_::8, shifted::bitstring>>), do: shifted
```

Comment:
There's a bit too much bit shifting going on here. The buffering of 3 extra bits at the start gets wiped out by the shift at the end and can be done away with entirely. The final shift calculation can also be simplified.

The decode function takes a 26 character Crockford encoded string and turns each character into a 5 bit chunk.

The end result is a 130 bit string, but since a valid Crockford encoded string is only 128 bits, we have to shift back 2 to get the proper length. According to the specs, this extra 2 bits comes off the leading character (with "7ZZZZZZZZZZZZZZZZZZZZZZZZZ" being the largest Crockford encoded string), so this can simply turn into:
```
  defp decode_bytes(<<>>, <<_::2, shifted::bitstring>>), do: shifted

  defp decode_bytes(<<codepoint::unsigned, rest::bitstring>>, acc) do
    decode_bytes(rest, <<acc::bitstring, i(codepoint)::unsigned-size(5)>>)
  end
```